### PR TITLE
CPR products generation code

### PR DESCRIPTION
**Pull request by leonardolaiolo**

### CODE REVIEW

**Comment by leonardolaiolo** [comment]
@mhidas here the pipeline config file for the CPR products. I did not push the file to the `octopus` branch because it is not an open pull request or draft, so cannot contribute with an extra commit, unless I open a draft pull request that I'm not sure it's your plan.
```
    "id": "IMOS_CPR_DB",
    "path": [ "IMOS_CPR_DB" ],
    "handler": "ImosBgcDbHandler",
    "params": {
        "include_regexes": [
            ".*\\.csv$"
        ],
        "harvest_params": {
            "db_schema": "imos_cpr_db",
            "db_objects": [
                {
                    "name": "cpr_samp",
                    "type": "table"
                },
                {
                    "name": "cpr_phyto_raw",
                    "type": "table"
                },
                {
                    "name": "cpr_phyto_changelog",
                    "type": "table"
                },
                {
                    "name": "cpr_zoop_raw",
                    "type": "table"
                },
                {
                    "name": "cpr_zoop_changelog",
                    "type": "table"
                },
                {
                    "name": "cpr_phytoplankton_map",
                    "type": "materialized view",
                    "dependencies": [
			"cpr_samp",
		        "cpr_phyto_raw"]
                },
                {
                    "name": "cpr_phytoplankton_abundance_raw_data",
                    "type": "materialized view",
                    "dependencies": ["cpr_phyto_raw"]
                },
                {
                    "name": "cpr_phytoplankton_abundance_species_data",
                    "type": "materialized view",
                    "dependencies": [
			"cpr_phyto_raw",
			"cpr_phyto_changelog"]
                },
                {
                    "name": "cpr_phytoplankton_abundance_genus_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_phyto_raw",
                        "cpr_phyto_changelog"]
                },
                {
                    "name": "cpr_phytoplankton_abundance_htg_data",
                    "type": "materialized view",
                    "dependencies": ["cpr_phyto_raw"]
                },
                {
                    "name": "cpr_zooplankton_map",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_samp",
                        "cpr_zoop_raw"]
                },
                {
                    "name": "cpr_zooplankton_abundance_raw_data",
                    "type": "materialized view",
                    "dependencies": ["cpr_zoop_raw"]
                },
                {
                    "name": "cpr_zooplankton_abundance_genus_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_zoop_raw",
                        "cpr_zoop_changelog"]
                },
                {
                    "name": "cpr_zooplankton_abundance_htg_data",
                    "type": "materialized view",
                    "dependencies": ["cpr_zoop_raw"]
                },
                {
                    "name": "cpr_zooplankton_abundance_copepods_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_samp",
                        "cpr_zoop_raw",
                        "cpr_zoop_changelog"
                    ]
                },
                {
                    "name": "cpr_zooplankton_abundance_non_copepods_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_samp",
                        "cpr_zoop_raw",
                        "cpr_zoop_changelog"
                    ]
                },
                {
                    "name": "cpr_phytoplankton_biovolume_raw_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_samp",
                        "cpr_phyto_raw"
                    ]
                },
                {
                    "name": "cpr_phytoplankton_biovolume_species_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_samp",
                        "cpr_phyto_raw",
                        "cpr_phyto_changelog"
                    ]
                },
                {
                    "name": "cpr_phytoplankton_biovolume_genus_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_samp",
                        "cpr_phyto_raw",
                        "cpr_phyto_changelog"
                    ]
                },
                {
                    "name": "cpr_phytoplankton_biovolume_htg_data",
                    "type": "materialized view",
                    "dependencies": [
                        "cpr_samp",
                        "cpr_phyto_raw"
                    ]
                }
	    ]
        }
    }
}
```

File: cpr_data/cpr_zooplankton_map.sql [hunk 1]
```diff
@@ -0,0 +1,25 @@
+-- This view is the basis for the WMS layer (seen on step 2 on AODN Portal).
+-- It also provides the metadata columns for all the zooplankton products.
+CREATE MATERIALIZED VIEW cpr_zooplankton_map AS
+  SELECT
+    s.trip_code AS "TripCode",
+    s.region AS "Region",
+    s.latitude AS "Latitude",
+    s.longitude AS "Longitude",
+    s.sampledateutc AT TIME ZONE 'UTC' AS "SampleDate_UTC",
```
Comment (mhidas) [inline cpr_data/cpr_zooplankton_map.sql:9]: Don't need `AT TIME ZONE 'UTC'` - same as in [`bgc_trip_metadata`](https://github.com/aodn/public-schema/blob/c6e65f83fbbb7f64a88b78f5582b87dc28c04676/bgc_data/bgc_trip_metadata.sql#L10)

**Review by mhidas — CHANGES_REQUESTED** [review]
I haven't done a full review, just looking at timestamp formatting...

File: cpr_data/cpr_phytoplankton_map.sql [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    s.trip_code AS "TripCode",
+    s.region AS "Region",
+    s.latitude AS "Latitude",
+    s.longitude AS "Longitude",
+    s.sampledateutc AT TIME ZONE 'UTC' AS "SampleDate_UTC",
+    s.sampledatelocal::text AS "SampleDate_Local", 
+    extract(year from s.sampledateutc)::int AS "Year_UTC",
+    extract(month from s.sampledateutc)::int AS "Month_UTC",
+    extract(day from s.sampledateutc)::int AS "Day_UTC",
+    to_char(sampledateutc, 'HH24:MI') AS "Time_UTC24hr",
```
Comment (mhidas) [inline cpr_data/cpr_phytoplankton_map.sql:14]: According to the product spec, these should all be "_local", not UTC.

File: cpr_data/cpr_phytoplankton_map.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    NULL AS "SatSST_degC", --to be updated
+    NULL AS "SatChlaSurf_mgm3", --to be updated
+    s.pci AS "PCI",
+    trip_code,
+    sample,
+    st_geomfromtext('POINT(' || longitude::text || ' ' || latitude::text || ')', 4326) AS geom
+    FROM cpr_samp s
+    WHERE sampletype LIKE '%P%' AND
+          trip_code IN (SELECT DISTINCT trip_code from cpr_phyto_raw)
```
Comment (mhidas) [inline cpr_data/cpr_phytoplankton_map.sql:23]: This needs last condition needs to use the `sample` column, which is what actually links the `cpr_phyto_raw` table back to `cpr_samp` (`trip_code` doesn't even exist in the former - the only reason this subquery doesn't break is it's interpreted to mean `cpr_samp.trip_code`, which makes this whole condition equivalent to `TRUE`)
```suggestion
          sample IN (SELECT DISTINCT sample from cpr_phyto_raw)
```

**Comment by mhidas** [comment]
I've fixed the couple of issues I mentioned in my recent comments.

Here's a complete set of products generated by the current code:
[cpr_products.zip](https://github.com/aodn/public-schema/files/8195489/cpr_products.zip)

**Comment by clairedavies** [comment]
this all looks good to me.
the only issue is with some change log information that I need to decide how to deal with, I can make these changes and they will flow through automatically.

**Review by mhidas — APPROVED** [review]

**Pull request merged**